### ISSUE TITLE
[Enhancement]: HPOS Adding/Deleting orders based on role

### ISSUE THREAD

**Issue by jitman** [woocommerce/woocommerce#39289]
### Describe the solution you'd like

We would like to prevent adding/deleting orders based on user role.

### Describe alternatives you've considered

Lot's of jQuery removal of DOM elements.

### Additional context

We have a highly customized order workflow and user roles.  Not all employees have the rights to delete orders and no one can create a 'blank' order via admin; our orders can only be created by choosing a customer to assign first.

**Comment by jorgeatorres** [comment]
@jitman: Thanks for the report and apologies for the delay here. HPOS on the admin still relies on the same capabilities (such as "edit post" or "publish") that the CPT implementation supported. Is this not working for you? In case this is still relevant to you, could you expand a bit on your use case and how HPOS works different than CPT in this regard?

**Comment by jitman** [comment]
Certainly!  I just double-checked if this was still an issue for me, and it is.  Prior to HPOS, I simply removed specific capabilities from the shop manger role.
```
$shop_manager = get_role( 'shop_manager' );
$shop_manager->remove_cap( 'publish_shop_orders' );
$shop_manager->remove_cap( 'delete_shop_order' );
$shop_manager->remove_cap( 'delete_shop_orders' );
```

This didn't work with HPOS enabled, so I tried hooking into `user_has_cap` to remove those capabilities prior to WP/WC checking a user's access:

```
add_filter( 'user_has_cap',function ( $allcaps, $caps, $args, $user ): array {

	unset( $allcaps['publish_shop_orders'] );
	$user->remove_cap( 'publish_shop_orders' );

	if ( ! in_array( 'administrator', $user->roles ) ) {
		unset( $allcaps['delete_shop_order'] );
		unset( $allcaps['delete_shop_orders'] );
		$user->remove_cap( 'delete_shop_order' );
		$user->remove_cap( 'delete_shop_orders' );
	}

	return $allcaps;
}, 999, 4 );
```

This also didn't work.  I ended up adding a little JS to remove the button after the page loads.

```
add_filter( 'views_woocommerce_page_wc-orders', function( array $views ): array {
	?>
    <script>
        jQuery(document).ready(function () {
            jQuery('a.page-title-action').text('Add order').remove();
        });
    </script>
	<?php
	return $views;
} );
```

**Comment by jorgeatorres** [comment]
Thanks for clarifying @jitman! We'll look into this.

**Comment by flakas** [comment]
Hi @jitman - thanks for raising this.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

It looks like this might be covered by the following work: https://github.com/woocommerce/woocommerce/pull/65528. I've linked your issue to it so you can follow the progress there.

Could you take a look and confirm whether that covers what you're reporting? If it doesn't quite match, let me know and we'll dig back in.

Where it's at: Keep the issue linked to PR #65528 and let engineering/PM confirm whether any non-delete capability gaps remain after that PR lands.

**Comment by jitman** [comment]
Looks like my original issue is being addressed between #58432 and #65528.  Thanks for following up!
